### PR TITLE
[Sol->Yul] Implement getters.

### PR DESCRIPTION
**Pull request by chriseth**
This should implement everything needed for all combinations of getters.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 1]
```diff
@@ -268,101 +268,134 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
+		code += Whiskers(R"(
+			let slot := <slot>
+			let offset := <offset>
+		)")
+		("slot", location.first.str())
+		("offset", to_string(location.second))
+		.render();
+
+		if (!paramTypes.empty())
+			solAssert(location.second == 0, "Zero offset expected if there are parameters.");
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp]: I'm not sure I understand this

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 2]
```diff
@@ -268,101 +268,134 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
+		TypePointer returnType = _varDecl.annotation().type;
+
+		vector<string> parameters;
+		vector<string> returnVariables;
+
+		for (size_t i = 0; i < paramTypes.size(); ++i)
+		{
+			MappingType const* mappingType = dynamic_cast<MappingType const*>(returnType);
+			ArrayType const* arrayType = dynamic_cast<ArrayType const*>(returnType);
+			solAssert(mappingType || arrayType, "");
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp:336]: Why are these inside the loop?

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 3]
```diff
@@ -268,101 +268,134 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
+
+			vector<string> keys = IRVariable("key_" + to_string(i),
+				mappingType ? *mappingType->keyType() : *TypeProvider::uint256()
+			).stackSlots();
+			parameters += keys;
+			code += Whiskers(R"(
+				slot := <indexAccess>(slot<?+keys>, <keys></+keys>)
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Just to confirm: is the idea to overwrite `slot`?

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 4]
```diff
@@ -268,101 +268,134 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
+			(
+				"indexAccess",
+				mappingType ?
+				m_utils.mappingIndexAccessFunction(*mappingType, *mappingType->keyType()) :
+				m_utils.storageArrayIndexAccessFunction(*arrayType)
+			)
+			("keys", joinHumanReadable(keys))
+			.render();
+
+			returnType = mappingType ? mappingType->valueType() : arrayType->baseType();
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Why is `returnType` initialized at declaration and reset here?

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp:336]: Accessors have the following form:
```
function accessor(uint a, uint b, uint c) returns (uint) {
  return x[a][b][c];
}
```
This is for e.g. `mapping(uint => mapping(uint => uint[]))`. Each loop iteration applies one index access.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Yes.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp]: I'll add some comments.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 5]
```diff
@@ -268,101 +268,145 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
+			solAssert(
+				location.second == 0,
+				"If there are parameters, we are dealing with structs or mappings and thus should have offset zero."
+			);
+
+		// The code of an accessor is of the form `x[a][b][c]` (it is slightly more complicated
+		// if the final type is a struct).
+		// In each iteration of the loop below, we consume one parameter, perform an
+		// index access, reassign the yul variable `slot` and move @a currentType further "down".
+		// The initial value of @a currentType is only used if we skipe the loop completely.
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGenerator.cpp]:
```suggestion
		// The initial value of @a currentType is only used if we skip the loop completely.
```

**Review by ekpyron — CHANGES_REQUESTED** [review]
Apart from the typo this is missing test cases for structs in general, for mappings to structs and arrays of structs, as well as for arrays of mappings and mappings of arrays, isn't it?

**Comment by chriseth** [comment]
It does not have tests mostly because other features need to be implemented first to be able to test structs and mappings (see the next PR in the row, for example).

**Comment by chriseth** [comment]
Added some more tests to https://github.com/ethereum/solidity/pull/9058

**Review by ekpyron — APPROVED** [review]

**Comment by ekpyron** [comment]
Let it be noted that I said that we should still have more tests for this, even beyond #9058 :-). Mappings to arrays of structs - arrays of mappings to structs, etc. - at least one for each combination would be nice.

**Pull request merged**